File: packages/runar-go/p384.go
// P-384 (NIST P-384 / secp384r1) off-chain helpers for testing Rúnar contracts.
//
// These functions use Go's standard library crypto/ecdsa and crypto/elliptic
// with no external dependencies. They are not compiled into Bitcoin Script —
// they exist so Go contract tests can generate keys, sign messages, and
// perform EC arithmetic using the P-384 curve.
//
// P384Point is a 96-byte ByteString: x[48] || y[48], big-endian, zero-padded.
// Coordinates are 48 bytes each (384 bits).
package runar

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"math/big"
)

// P384Point is a 96-byte ByteString encoding a P-384 point as x[48] || y[48].
// Coordinates are big-endian and zero-padded to 48 bytes each.
type P384Point = ByteString

// P384KeyPair holds a P-384 key pair.
type P384KeyPair struct {
	SK           *ecdsa.PrivateKey
	PK           P384Point // 96-byte uncompressed encoding (x[48] || y[48])
	PKCompressed []byte    // 49-byte compressed encoding (02/03 prefix + x[48])
}

// p384PointFromXY serializes a P-384 (x, y) pair into a 96-byte P384Point.
func p384PointFromXY(x, y *big.Int) P384Point {
	buf := make([]byte, 96)
	xb := x.Bytes()
	yb := y.Bytes()
	copy(buf[48-len(xb):48], xb)
	copy(buf[96-len(yb):96], yb)
	return P384Point(buf)
}

// p384XYFromPoint extracts (x, y) *big.Int from a 96-byte P384Point.
func p384XYFromPoint(p P384Point) (*big.Int, *big.Int) {
	b := []byte(p)
	if len(b) != 96 {
		panic("runar: P384Point must be exactly 96 bytes")
	}
	x := new(big.Int).SetBytes(b[:48])
	y := new(big.Int).SetBytes(b[48:])
	return x, y
}

// P384Keygen generates a random P-384 key pair.
func P384Keygen() P384KeyPair {
	sk, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
	if err != nil {
		panic("runar: P384Keygen: " + err.Error())
	}
	pk := p384PointFromXY(sk.PublicKey.X, sk.PublicKey.Y)
	compressed := elliptic.MarshalCompressed(elliptic.P384(), sk.PublicKey.X, sk.PublicKey.Y)
	return P384KeyPair{
		SK:           sk,
		PK:           pk,
		PKCompressed: compressed,
	}
}

// P384Sign signs msg with sk using ECDSA on P-384.
// The message is SHA-256 hashed internally before signing.
// Returns a 96-byte raw signature: r[48] || s[48] (big-endian, zero-padded).
func P384Sign(msg []byte, sk *ecdsa.PrivateKey) []byte {
	digest := sha256.Sum256(msg)
	r, s, err := ecdsa.Sign(rand.Reader, sk, digest[:])
	if err != nil {
		panic("runar: P384Sign: " + err.Error())
	}
	sig := make([]byte, 96)
	r.FillBytes(sig[:48])
	s.FillBytes(sig[48:])
	return sig
}

// P384Add adds two P-384 points.
// If either point is the zero point (all-zero bytes), the other is returned.
func P384Add(a, b P384Point) P384Point {
	ax, ay := p384XYFromPoint(a)
	bx, by := p384XYFromPoint(b)
	curve := elliptic.P384()
	rx, ry := curve.Add(ax, ay, bx, by)
	return p384PointFromXY(rx, ry)
}

// P384Mul performs scalar multiplication k * p on the P-384 curve.
func P384Mul(p P384Point, k *big.Int) P384Point {
	px, py := p384XYFromPoint(p)
	curve := elliptic.P384()
	kBytes := k.Bytes()
	rx, ry := curve.ScalarMult(px, py, kBytes)
	return p384PointFromXY(rx, ry)
}

// P384MulGen performs scalar multiplication k * G where G is the P-384 generator.
func P384MulGen(k *big.Int) P384Point {
	curve := elliptic.P384()
	kBytes := k.Bytes()
	rx, ry := curve.ScalarBaseMult(kBytes)
	return p384PointFromXY(rx, ry)
}

// P384Negate returns the negation of a P-384 point: (x, p - y).
func P384Negate(p P384Point) P384Point {
	px, py := p384XYFromPoint(p)
	fieldP := elliptic.P384().Params().P
	negY := new(big.Int).Sub(fieldP, py)
	negY.Mod(negY, fieldP)
	return p384PointFromXY(px, negY)
}

// P384OnCurve reports whether the given point lies on the P-384 curve.
func P384OnCurve(p P384Point) bool {
	px, py := p384XYFromPoint(p)
	// Reject the point at infinity (0, 0) — not a valid curve point.
	if px.Sign() == 0 && py.Sign() == 0 {
		return false
	}
	return elliptic.P384().IsOnCurve(px, py)
}

// P384EncodeCompressed returns the 49-byte compressed encoding of a P-384 point
// (02/03 prefix byte followed by the 48-byte x coordinate, big-endian).
func P384EncodeCompressed(p P384Point) ByteString {
	px, py := p384XYFromPoint(p)
	compressed := elliptic.MarshalCompressed(elliptic.P384(), px, py)
	return ByteString(compressed)
}

// VerifyECDSAP384 verifies a raw ECDSA signature on P-384.
// msg is the raw message — it is SHA-256 hashed internally before verification.
// sig is the 96-byte raw signature (r[48] || s[48]) produced by P384Sign.
// pubkey is the 49-byte compressed public key (02/03 + x[48]).
func VerifyECDSAP384(msg, sig, pubkey ByteString) bool {
	if len(sig) != 96 {
		return false
	}
	digest := sha256.Sum256([]byte(msg))

	// Decode compressed public key.
	x, y := elliptic.UnmarshalCompressed(elliptic.P384(), []byte(pubkey))
	if x == nil {
		return false
	}
	pub := &ecdsa.PublicKey{Curve: elliptic.P384(), X: x, Y: y}

	r := new(big.Int).SetBytes([]byte(sig)[:48])
	s := new(big.Int).SetBytes([]byte(sig)[48:])
	return ecdsa.Verify(pub, digest[:], r, s)
}

File: packages/runar-rs/src/p384.rs
//! P-384 (NIST P-384 / secp384r1) off-chain helpers for testing Rúnar contracts.
//!
//! These functions use the `p384` crate for real P-384 operations.
//! They are not compiled into Bitcoin Script — they exist so Rust contract tests
//! can generate keys, sign messages, and verify P-384 signatures.
//!
//! `P384KeyPair` holds a random P-384 key pair. `p384_sign` signs a message
//! (SHA-256 hashed internally, matching the on-chain codegen which uses OP_SHA256
//! for both P-256 and P-384). `verify_ecdsa_p384` verifies a raw r||s signature
//! against a 49-byte compressed public key.
//!
//! Note: P-384 ECDSA normally uses SHA-384. The on-chain codegen uses OP_SHA256
//! for both P-256 and P-384 for script-size reasons, so these helpers mirror
//! that choice by hashing with SHA-256 and calling the prehash signing API.

use p384::ecdsa::{signature::hazmat::{PrehashSigner, PrehashVerifier}, SigningKey, VerifyingKey, Signature};
use sha2::{Digest, Sha256};
use rand::rngs::OsRng;

/// A P-384 key pair.
pub struct P384KeyPair {
    sk: SigningKey,
    /// 96-byte uncompressed public key encoding: x[48] || y[48].
    pub pk: Vec<u8>,
    /// 49-byte compressed public key: (02/03) prefix + x[48].
    pub pk_compressed: Vec<u8>,
}

/// Generate a random P-384 key pair.
pub fn p384_keygen() -> P384KeyPair {
    let sk = SigningKey::random(&mut OsRng);
    let vk = sk.verifying_key();
    let point = vk.to_encoded_point(false); // uncompressed
    let coords = point.as_bytes();
    // uncompressed: 0x04 || x[48] || y[48] → strip prefix
    let pk = coords[1..].to_vec(); // 96 bytes

    let compressed_point = vk.to_encoded_point(true);
    let pk_compressed = compressed_point.as_bytes().to_vec();

    P384KeyPair { sk, pk, pk_compressed }
}

/// Sign `msg` with P-384 ECDSA. The message is SHA-256 hashed internally
/// (matching the on-chain codegen which uses OP_SHA256 for both curves).
/// Returns a 96-byte raw signature: r[48] || s[48] (big-endian, zero-padded).
pub fn p384_sign(msg: &[u8], kp: &P384KeyPair) -> Vec<u8> {
    let digest = Sha256::digest(msg);
    let sig: Signature = kp.sk.sign_prehash(&digest)
        .expect("p384_sign: sign_prehash failed");
    let bytes = sig.to_bytes();
    bytes.to_vec()
}

/// Verify a P-384 ECDSA signature.
///
/// - `msg`: raw message (SHA-256 hashed internally).
/// - `sig`: 96-byte raw signature r[48] || s[48].
/// - `pubkey`: 49-byte compressed P-384 public key (02/03 prefix + x[48]).
pub fn verify_ecdsa_p384(msg: &[u8], sig: &[u8], pubkey: &[u8]) -> bool {
    if sig.len() != 96 {
        return false;
    }
    let vk = match VerifyingKey::from_sec1_bytes(pubkey) {
        Ok(k) => k,
        Err(_) => return false,
    };
    let sig = match Signature::from_slice(sig) {
        Ok(s) => s,
        Err(_) => return false,
    };
    let digest = Sha256::digest(msg);
    vk.verify_prehash(&digest, &sig).is_ok()
}
